### ISSUE TITLE
Design Review 2018-10-18 00:00 UTC (Asia/Oceania)

### ISSUE THREAD

**Issue by mrjoro** [ampproject/amphtml#17925]
Time: [2018-10-18 00:00 UTC](https://www.timeanddate.com/worldclock/meeting.html?year=2018&month=10&day=18&iv=0) ([add to Google Calendar](http://www.google.com/calendar/event?action=TEMPLATE&text=AMP%20Project%20Design%20Review&dates=20181018T000000Z/20181018T010000Z&details=https%3A%2F%2Fhangouts.google.com%2Fhangouts%2F_%2Fgoogle.com%2Famp-eng-desrev))
Location: [Video conference via Google Hangouts](https://hangouts.google.com/hangouts/_/google.com/amp-eng-desrev)

The AMP Project holds weekly engineering [design reviews](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md).  **We encourage everyone in the community to participate in these design reviews.**

If you are interested in bringing your design to design review, read the [design review documentation](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md) and add a link to your design doc by the Monday before your design review.

When attending a design review please read through the designs *before* the design review starts.  This allows us to spend more time on discussion of the design.

We rotate our design review between times that work better for different parts of the world as described in our [design review documentation](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md), but you are welcome to attend any design review.  If you cannot make any of the design reviews but have a design to discuss please let mrjoro@ know on [Slack](https://github.com/ampproject/amphtml/blob/master/CONTRIBUTING.md#discussion-channels) and we will find a time that works for you.

**Comment by cathyxz** [comment]
# AMP List Resize
Moved here from 11-14-2018. 

## TLDR
We would like an option to allow `<amp-list>` to resize on user interaction without violating AMP principles (e.g. no resizing if bottom is in viewport). Feel free to skip to the "Potential Fixes" section below. 

## Goals
It needs to resize when: 
1. `<amp-list>` `src` is changed via `<amp-bind>`. 
2.  `<amp-list>` contents change size due to valid user interaction INSIDE `<amp-list>`, e.g. `<amp-list>` contains `<amp-accordion>`, which the user toggles to expand / collapse. 
3. `<amp-list>` contents change size due to valid user interaction OUTSIDE `<amp-list>`, e.g. styling on `<amp-list>`'s children changes due to clicking on a button that toggles styles via `<amp-bind>`. 
4. ([Already handled] Users click on the overflow element, e.g. loaded contents size > `<amp-list>` initial size. )

## Strategy
First strategy implemented in https://github.com/ampproject/amphtml/pull/16564 adds an `auto-resize` attribute which resets the layout to `CONTAINER` on first load. This handles the 3 new cases above, and overrides / bypasses case 4 completely. 

## Problem
Changing to layout `CONTAINER` "on first load" violates the "no resizing if bottom is in viewport" principle. 

## Fix 
Only change to layout `CONTAINER` after the overflow attribute has already been clicked per https://github.com/ampproject/amphtml/pull/18543. This will avoid violating the "no resizing if bottom is in viewport" principle. It also successfully handles case 1.

## Problem
It does NOT handle case 2 and 3, since by default, if the content in `<amp-list>` is initially sized correctly, further size updates will not result in an `attemptChangeSize`, will not show the overflow button, and will not give an opportunity to change to layout `CONTAINER`. 

## Potential Fixes
### Resize Action
Introduce a `resize` action that is basically an action that changes to layout `CONTAINER`. This would allow developers to call it manually on actions like `tap` or `<amp-accordion>`'s `expand` and `collapse`. Pro: it allows developers to manually deal with case 2 and 3 above. Con: requires developer action. 

### Mutation Observer or Tap Listener
1. Use a tap listener on `<amp-list>` to detect user interactions. If `auto-resize`, change to layout `CONTAINER` on tap. 
2. Use a MutationObserver on `<amp-list>` to detect changes to the contents of `<amp-list>`. If `auto-resize`, change to layout `CONTAINER` when a mutation occurs. 
3. Use a tap listener on the page AND a MutationObserver. If `auto-resize`, change to layout `CONTAINER` when a tap results in a mutation. 

Pros: no developer action required. 
Cons: mutation observer could be expensive, tap listener only could be too imprecise, there might be race conditions around using both. 
Note: the action on tap / on mutation could be either call attemptChangeSize or directly call the changeToLayoutContainer function. 

### Other strategies??? 
Open to feedback here. 

## Context
Original issue here: https://github.com/ampproject/amphtml/issues/10230
Current issue here: https://github.com/ampproject/amphtml/issues/18376
Initial PR here: https://github.com/ampproject/amphtml/pull/16564
Update PR here: https://github.com/ampproject/amphtml/pull/18543

/cc @choumx @aghassemi

**Comment by cathyxz** [comment]
Okay since the people that I need to discuss this with can't make it to this design review, I'm going to move it again to a later date.

**Comment by mrjoro** [comment]
Cancelling since there are no topics.
